src/cluster_resources.rs

```diff
@@ -0,0 +1,393 @@
+    /// instance but are not contained in the given list.
+    ///
+    /// If it is forbidden to list the resources of the given kind then it is assumed that the
+    /// caller is not in charge of these resources, the deletion is skipped, and no error is
+    /// returned.
+    ///
+    /// # Arguments
+    ///
+    /// * `client` - The client which is used to access Kubernetes
+    async fn delete_orphaned_resources<T: ClusterResource>(
+        &self,
+        client: &Client,
+    ) -> OperatorResult<()> {
+        match self.list_deployed_cluster_resources::<T>(client).await {
+            Ok(deployed_cluster_resources) => {
+                let orphaned_resources = deployed_cluster_resources
+                    .into_iter()
+                    .filter(|r| !self.resources.contains(&r.into()))
+                    .collect::<Vec<_>>();
+
+                if !orphaned_resources.is_empty() {
+                    info!(
+                        "Deleting orphaned {}: {}",
+                        T::plural(&()),
+                        ClusterResources::print_resources(&orphaned_resources),
+                    );
+                    for resource in orphaned_resources.iter() {
+                        client.delete(resource).await?;
+                    }
+                }
+
+                Ok(())
+            }
+            Err(Error::KubeError {
+                source: kube::Error::Api(ErrorResponse { code: 403, .. }),
+            }) => {
+                debug!(
+                    "Skipping deletion of orphaned {} because the operator is not allowed to list \
+                      them and is therefore probably not in charge of them.",
+                    T::plural(&())
+                );
+                Ok(())
+            }
+            Err(error) => Err(error),
+        }
+    }
+
+    /// Creates a string containing the names and if present namespaces of the given resources
+    /// sorted by name and separated with commas.
+    fn print_resources<T: ClusterResource>(resources: &[T]) -> String {
+        let mut output = resources
+            .iter()
+            .map(ClusterResources::print_resource)
+            .collect::<Vec<_>>();
+        output.sort();
+        output.join(", ")
+    }
+
+    /// Creates a string containing the name and if present namespace of the given resource.
+    fn print_resource<T: ClusterResource>(resource: &T) -> String {
+        if let Some(namespace) = resource.namespace() {
+            format!("{name}.{namespace}", name = resource.name())
+        } else {
+            resource.name()
+        }
+    }
+
+    /// Lists the deployed resources with instance, name, and managed-by labels equal to this
+    /// cluster instance.
+    ///
+    /// # Arguments
+    ///
+    /// * `client` - The client which is used to access Kubernetes
+    async fn list_deployed_cluster_resources<T: ClusterResource>(
```

Comment:
Should this maybe be 'pub'?

Reply:
This function lists all resources which are labelled if they belong to this cluster instance. This list can contain orphaned resources which were not added to `ClusterResources` before. It would confuse the user of the `ClusterResources` struct if this function would be public:

```rust
let mut cluster_resources = ClusterResources::new(...);

// deploys resource1
cluster_resources.add(client, resource1).await?;

// returns resource1 and resource2
let deployed_cluster_resources = cluster_resources.list_deployed_cluster_resources(client).await?;

// Why do the cluster_resources differ from the deployed_cluster_resources?
```

Reply:
ok

Reply:
Actually I have spent some more thought on this..

I am currently in need of a fn sort of like this one for the NiFi operator. The scenario is an update of the NiFi version, which requires a full restart of the cluster. So the reconciliation checks fairly early on, whether the version is different and then scales all deployed statefulsets to 0.
For this I'd like to list the deployed statefulsets for this cluster and iterate over these, which could be done one of two ways: 
- list the resources in k8s
- generate the list of statefulsets in the regular reconciliation 

I think there is no right or wrong way here, it is more a question of "how do we want to treat concurrent edits of the NifiCluster resource" - so basically someone changes the version and scales the cluster in the same edit of the NifiCluster object.

It would probably make more sense to do the scaling first to avoid upgrading nodes just to then delete them, but from an implementation point of view I think doing the upgrade first might be easier. And then it would actually make sense to be able to list the owned statefulsets for this cluster.

Reply:
I looked at the `feat/update` branch of the NiFi operator and several questions arose. Perhaps we should have a meeting and present only the outcome in this discussion thread?